### PR TITLE
Generate random configurations of fragment tapes for sample circuit cutting

### PR DESCRIPTION
**Pull request by anthayes92**
**Context:**
When cutting a circuit containing sample measurements, it is necessary to generate configurations by replacing `MeasureNodes` with randomly sampled measurements from the available basis and replacing `PrepareNodes` with state preparations that are sampled from the corresponding pair of eigenvectors. 

**Description of the Change:**
A function has been added to generate random configurations of fragment tapes.

**Benefits:**
Facilitates the circuit cutting pipeline for sampling measurements 

**Possible Drawbacks:**
Potential computational bottleneck depending on number of shots requested for execution.

### CODE REVIEW

**Comment by anthayes92** [comment]
[sc-15880]

File: pennylane/transforms/qcut.py [hunk 1]
```diff
@@ -649,6 +702,125 @@ def expand_fragment_tape(
...
+
+    >>> g = qcut.tape_to_graph(tape)
+    >>> qcut.replace_wire_cut_nodes(g)
+    >>> subgraphs, communication_graph = qcut.fragment_graph(g)
+    >>> tapes = [qcut.graph_to_tape(sg) for sg in subgraphs]
+
+    We can then expand over the measurement and preparation nodes to generate random
+    configurations using:
+
+    .. code-block:: python
```
Comment (anthayes92) [inline pennylane/transforms/qcut.py:751]: code block needed for correct rendering of diagrams below

File: pennylane/transforms/qcut.py [hunk 2]
```diff
@@ -633,6 +700,126 @@ def expand_fragment_tape(
...
+
+        >>> configs = qcut.expand_fragment_tapes_mc(tapes, communication_graph, 3)
+        >>> i = 0
```
Comment (trbromley) [inline pennylane/transforms/qcut.py:752]:
```suggestion
```

**Review by trbromley — COMMENTED** [review]
Thanks @anthayes92, left some comments for now

File: tests/transforms/test_qcut.py [hunk 3]
```diff
@@ -1582,6 +1722,233 @@ def test_no_measure_node_observables(self):
             compare_tapes(tape, config)
 
 
+class TestExpandFragmentTapesMC:
+    """
+    Tests fragment tapes are correctly expanded to all random configurations
```
Comment (trbromley) [inline tests/transforms/test_qcut.py:1727]:
```suggestion
    Tests that fragment tapes are correctly expanded to all random configurations
```

File: tests/transforms/test_qcut.py [hunk 4]
```diff
@@ -1582,6 +1722,233 @@ def test_no_measure_node_observables(self):
...
+    for the Monte Carlo sampling technique.
+    """
+
+    def test_expand_mc(self, monkeypatch):
+        """
+        Tests that fragment configurations are generated correctly using the
+        `expand_fragment_tapes_mc` method.
```
Comment (trbromley) [inline tests/transforms/test_qcut.py:1734]:
```suggestion
        `expand_fragment_tapes_mc` function.
```

File: tests/transforms/test_qcut.py [hunk 5]
```diff
@@ -1582,6 +1722,233 @@ def test_no_measure_node_observables(self):
...
+            qml.sample(qml.Projector([1], wires=[2]))
+
+        tapes = [tape0, tape1]
+
+        edge_data = {"pair": (tape0.operations[2], tape1.operations[0])}
+        communication_graph = MultiDiGraph([(0, 1, edge_data)])
+
+        with monkeypatch.context() as m:
+            m.setattr(
+                np.random, "choice", lambda a, size=(1, 3), replace=True: np.array([[4, 0, 1]])
```
Comment (trbromley) [inline tests/transforms/test_qcut.py:1755]:
```suggestion
                np.random, "choice", lambda a, size, replace: np.array([[4, 0, 1]])
```
Do we need the defaults?

File: tests/transforms/test_qcut.py [hunk 6]
```diff
@@ -1582,6 +1722,233 @@ def test_no_measure_node_observables(self):
...
+                for op in frag_1_ops_and_meas:
+                    qml.apply(op)
+            expected_tapes_1.append(expected_tape)
+
+        for tape, exp_tape in zip(fragment_configurations[1], expected_tapes_1):
+            compare_tapes(tape, exp_tape)
+
+    def test_expand_multinode_frag(self, monkeypatch):
+        """
+        Tests that a fragments with multiple measure and prepare nodes are
```
Comment (trbromley) [inline tests/transforms/test_qcut.py:1805]:
```suggestion
        Tests that fragments with multiple measure and prepare nodes are
```

File: tests/transforms/test_qcut.py [hunk 7]
```diff
@@ -1582,6 +1722,233 @@ def test_no_measure_node_observables(self):
...
+            (0, 1, {"pair": (tape0.operations[8], tape1.operations[4])}),
+            (1, 0, {"pair": (tape1.operations[6], tape0.operations[9])}),
+        ]
+        communication_graph = MultiDiGraph(edge_data)
+
+        with monkeypatch.context() as m:
+            m.setattr(
+                np.random,
+                "choice",
+                lambda a, size=(4, 2), replace=True: np.array([[4, 6], [1, 2], [2, 3], [3, 0]]),
```
Comment (trbromley) [inline tests/transforms/test_qcut.py:1848]:
```suggestion
                lambda a, size, replace: np.array([[4, 6], [1, 2], [2, 3], [3, 0]]),
```

**Review by trbromley — COMMENTED** [review]
Thanks @anthayes92, tests look good so far!

File: pennylane/transforms/qcut.py [hunk 8]
```diff
@@ -649,6 +700,125 @@ def expand_fragment_tape(
...
+        M0 =
+        [1]
+
+        1: ──X──H──S──╭C──┤ Sample[Projector(M0)]
+        2: ───────────╰X──┤ Sample[Projector(M0)]
+        M0 =
+        [1]
+    """
+    pairs = [e[-1] for e in communication_graph.edges.data("pair")]
+    settings = np.random.choice(range(8), size=(len(pairs), shots), replace=True)
```
[the remark this answers is not in the document]
Comment (anthayes92) [reply pennylane/transforms/qcut.py:795]: I tried this out and while it seems to give similar output to `np.random.choice` the monkeypatch in the tests doesn't hold. Really not sure what's happening here... 

For clarity I use: 
```python
        fixed_choice = np.array([[4, 6], [1, 2], [2, 3], [3, 0]])
        with monkeypatch.context() as m:
            m.setattr(
                qml.math.random,
                "choice",
                lambda a, size, replace: fixed_choice,
            )

```
which doesn't give any errors but just seems to be ignored!

File: tests/transforms/test_qcut.py (hunk 5 above)
Comment (anthayes92) [reply tests/transforms/test_qcut.py]: manually checked and added here: f9b8ddf4bc34bf64ce9b0d9743331d157abbf9b3

Thanks!

File: pennylane/transforms/qcut.py [hunk 9]
```diff
@@ -633,6 +700,126 @@ def expand_fragment_tape(
...
+                for meas in tape.measurements:
+                    qml.apply(meas)
+                for op in tape.operations:
+                    if isinstance(op, MeasureNode):
+                        MC_MEASUREMENTS[meas_settings[op.id][shot]](w)
+            frag_config.append(new_tape)
+
+        all_configs.append(frag_config)
+
+    return all_configs
```
[the remark this answers is not in the document]
Comment (anthayes92) [reply pennylane/transforms/qcut.py]: added here along with assertions in unit tests: f9b8ddf4bc34bf64ce9b0d9743331d157abbf9b3

File: pennylane/transforms/qcut.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (anthayes92) [reply pennylane/transforms/qcut.py:795]: I tried this out and while it seems to give similar output to `np.random.choice` the monkeypatch in the tests doesn't hold. Really not sure what's happening here... 

For clarity I use: 
```python
        fixed_choice = np.array([[4, 6], [1, 2], [2, 3], [3, 0]])
        with monkeypatch.context() as m:
            m.setattr(
                qml.math.random,
                "choice",
                lambda a, size, replace: fixed_choice,
            )

```
which doesn't give any errors but just seems to be ignored!

**Review by trbromley — APPROVED** [review]
Thanks @anthayes92, nice work!

File: pennylane/transforms/qcut.py [hunk 10]
```diff
@@ -649,6 +661,168 @@ def expand_fragment_tape(
...
+    _pauliY,
+    _pauliY,
+    _pauliZ,
+    _pauliZ,
+]
+
+
+def expand_fragment_tapes_mc(
+    tapes: Sequence[QuantumTape], communication_graph: MultiDiGraph, shots: int
+) -> List[QuantumTape]:
```
Comment (trbromley) [inline pennylane/transforms/qcut.py:706]:
```suggestion
) -> Tuple[List[QuantumTape], np.array]:
```

File: tests/transforms/test_qcut.py [hunk 11]
```diff
@@ -1722,6 +1722,241 @@ def test_no_measure_node_observables(self):
...
+        communication_graph = MultiDiGraph([(0, 1, edge_data)])
+
+        fixed_choice = np.array([[4, 0, 1]])
+        with monkeypatch.context() as m:
+            m.setattr(np.random, "choice", lambda a, size, replace: fixed_choice)
+            fragment_configurations, settings = qcut.expand_fragment_tapes_mc(
+                tapes, communication_graph, 3
+            )
+
+        assert np.isclose(settings, fixed_choice).all()
```
Comment (trbromley) [inline tests/transforms/test_qcut.py:1760]:
```suggestion
        assert np.allclose(settings, fixed_choice)
```

File: tests/transforms/test_qcut.py [hunk 12]
```diff
@@ -1722,6 +1722,241 @@ def test_no_measure_node_observables(self):
...
+            m.setattr(
+                np.random,
+                "choice",
+                lambda a, size, replace: fixed_choice,
+            )
+            fragment_configurations, settings = qcut.expand_fragment_tapes_mc(
+                tapes, communication_graph, 2
+            )
+
+        assert np.isclose(settings, fixed_choice).all()
```
Comment (trbromley) [inline tests/transforms/test_qcut.py:1858]:
```suggestion
        assert np.allclose(settings, fixed_choice)
```

**Pull request merged**